ownCloud/Client/Actions/Action.swift

```diff
@@ -0,0 +1,260 @@
+		return .middle
+	}
+
+	// MARK: - Finding actions
+	class func sortedApplicableActions(for context: ActionContext) -> [Action] {
+		var sortedActions : [Action] = []
+
+		if let matches = try? OCExtensionManager.shared.provideExtensions(for: context) {
+			for match in matches {
+				if let action = match.extension.provideObject(for: context) as? Action {
+					sortedActions.append(action)
+				}
+			}
+		}
+
+		sortedActions.sort { (action1, action2) -> Bool in
+			return action1.position.rawValue < action2.position.rawValue
+		}
+
+		return sortedActions
+	}
+
+	// MARK: - Provide Card view controller
+	class func cardViewController(for item: OCItem, with context: ActionContext, progressHandler: ((Progress) -> Void)? = nil, completionHandler: ((Error?)-> Void)? = nil) -> UIViewController {
+
+		let tableViewController = MoreStaticTableViewController(style: .grouped)
+		let header = MoreViewHeader(for: item, with: context.core!)
+		let moreViewController = MoreViewController(item: item, core: context.core!, header: header, viewController: tableViewController)
+
+		let title = NSAttributedString(string: "Actions", attributes: [NSAttributedStringKey.font: UIFont.systemFont(ofSize: 20, weight: .heavy)])
+
+		let actions = Action.sortedApplicableActions(for: context)
+
+		actions.forEach({
+			$0.actionWillRunHandler = {
+				moreViewController.dismiss(animated: true)
+			}
+
+			$0.progressHandler = progressHandler
+
+			$0.completionHandler = completionHandler
+		})
+
+		let actionsRows: [StaticTableViewRow] = actions.compactMap({return $0.provideStaticRow()})
+
+		tableViewController.addSection(MoreStaticTableViewSection(headerAttributedTitle: title, identifier: "actions-section", rows: actionsRows))
+
+		return moreViewController
+	}
+
+	// MARK: - Action metadata
+	var context : ActionContext
+	var actionExtension: ActionExtension
+	var core : OCCore
+
+	// MARK: - Action creation
+	required init(for actionExtension: ActionExtension, with context: ActionContext) {
+		self.actionExtension = actionExtension
+		self.context = context
+		self.core = context.core!
+
+		super.init()
+	}
+
+	// MARK: - Execution metadata
+	var progressHandler : ActionProgressHandler?     // to be filled before calling run(), provideStaticRow(), provideContextualAction(), etc. if desired
+	var completionHandler : ActionCompletionHandler? // to be filled before calling run(), provideStaticRow(), provideContextualAction(), etc. if desired
+	var actionWillRunHandler: ActionWillRunHandler? // to be filled before calling run(), provideStaticRow(), provideContextualAction(), etc. if desired
+
+	// MARK: - Action implementation
+	func willRun() {
+
+		if Thread.isMainThread == false {
+			Log.warning("The Run method of the action \(Action.identifier!.rawValue) is not called inside the main thread")
+		}
+
+		if actionWillRunHandler != nil {
+			actionWillRunHandler!()
+		}
+	}
+
+	func run() {
+		if completionHandler != nil {
+			completionHandler!(nil)
+		}
+	}
+
```

Comment:
**Completely optional**, but as I reviewed the code I thought this would make for an even nicer implementation overall (and may possibly also prove useful further down the road):
- add new `func completed(with: Error? = nil)` that calls the completionHandler (if one exists)
- throughout the actions, call `self.completed(error)` (or `self.completed()` where applicable) instead of the `completionHandler` directly.
- same for progressHandler with f.ex. a `func publish(progress: Progress)` that calls the `progressHandler` and replaces direct calls of it